schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,292 @@
+
+Each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links.
+
+## Structure
+
+### Parameters
+
+Configurable parameters for any given IPLD HashMap:
+
+* `hashAlg`: The hash algorithm applied to keys in order to evenly distribute entries throughout the data structure. The algorithm is chosen based on speed, byte-size and randomness properties (but it must be available to the reader, hence the need for shared defaults, see below).
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
```

Comment:
I originally did use the multihash varint for this but I got sick of the terseness and indirection for very little gain.
I could switch back if someone has strong feelings about this but my default preference is for these data structures to be _somewhat_ readable on their own—at for the debugging case, to avoid indirection as much as possible (both code indirection "what is this varint again?" and cognitive indirection, "what's `p` doing again?").

Reply:
It should be fine, we just need to make sure we don't change the names of these codecs. I wonder if we should come up with a namespaced naming scheme and do a mass-rename first.